Test/CoreSDK.Test/Net40/TaskExTests.cs

```diff
@@ -0,0 +1,173 @@
+namespace Microsoft.ApplicationInsights
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using Assert = Xunit.Assert;
+
+    [TestClass]
+    public class TaskExTests
+    {
+
+        [TestMethod]
+        public void RethrowIfFailedDoesntThrowIfNoExceptionOccured()
+        {
+            var task = Task.Factory.StartNew(() => { });
+
+            task.Wait();
+
+            Assert.DoesNotThrow(() => task.RethrowIfFailed());
+        }
+
+        [TestMethod]
+        public void RethrowIfFailedThrowsIfTaskIsNotCompleted()
+        {
+            var task = TaskEx.Delay(TimeSpan.FromMilliseconds(100));
+
+            Assert.Throws<ArgumentException>(() => task.RethrowIfFailed());
+        }
+
+        [TestMethod]
+        public void RethrowIfFailedThrowsIfExceptionOccured()
+        {
+            var task = Task.Factory.StartNew(() => { throw new Exception(); });
+
+            try
+            {
+                task.Wait();
+            }
+            catch (Exception)
+            {
+                // ignore
+            }
+
+            Assert.Throws<AggregateException>(() => task.RethrowIfFailed());
+        }
+
+        [TestMethod]
+        public void DelayWithZeroTimeoutCompletesTask()
+        {
+            var task = TaskEx.Delay(TimeSpan.Zero);
+            task.Wait();
+
+            Assert.Equal(task.Status, TaskStatus.RanToCompletion);
+        }
+
+        [TestMethod]
+        public void DelayWaitsAtLeastTimeout()
+        {
+            var timeout = TimeSpan.FromMilliseconds(50);
+            var time = DateTime.UtcNow;
+            var task = TaskEx.Delay(timeout);
+            
+            task.Wait();
+            var spent = DateTime.UtcNow - time;
+
+            Assert.True(spent >= timeout);
+        }
+
+        [TestMethod]
+        public void DelayWhichIsCanceledThrows()
+        {
+            var timeout = TimeSpan.FromSeconds(1);
+            var cancellationTokenSource = new CancellationTokenSource();
+            var task = TaskEx.Delay(timeout, cancellationTokenSource.Token);
+
+            cancellationTokenSource.Cancel();
+
+            Assert.Throws<AggregateException>(() => task.Wait());
```

Comment:
To be safe, the test should look into the AggregateException and make sure the cancellation is indeed the reason.